### ISSUE TITLE
periodic_box.prm fails

### ISSUE THREAD

**Issue by tjhei** [geodynamics/aspect#255]
see test output here: http://cdash.kyomu.43-1.org/testDetails.php?test=4826757&build=2653

The matrix A is not invertible because of the periodic boundary conditions. We already now that the coarse grid solver of AMG (set to "KLU") is the problem here.

This test started failing when we changed the order of the mapping from Q4 to Q1, even though we are operating on a box, so that shouldn't matter. See #242 for a lengthy discussion without a solution so far.

**Comment by bangerth** [comment]
Isn't it correct, though, that in that case there matrix does and should have a null space? Do we need to impose constraints for the testcase to remove the nullspace?

**Comment by tjhei** [comment]
What we have done successfully in the past is to:
1. ignore the nullspace during assembly/solve with GMRES staying in the same subspace
2. remove the nullspace as a postprocess

**Comment by bangerth** [comment]
Good point. So what do we do here?

**Comment by tjhei** [comment]
Before we mark this as resolved, I have one more data point:

Take tests/depth_average_01.prm, increase global refinement by one and switch back to Q4 -> crash. This can be fixed by either switching to Q1 or by setting cheap solver steps = 0. Crazy.

**Comment by bangerth** [comment]
The way I read the discussion over on the deal.II testcase I created is that this is really a failure in Trilinos to compute things relative, not absolute. I think we need to do something like this in the Trilinos solver wrappers:
```
  const double scaling_factor = 1./rhs.l2_norm();
  rhs *= scaling_factor;
  solution *= scaling_factor;
  ... do as before...
  rhs /= scaling_factor;
  solution /= scaling_factor;
```  
I'll have to try something along these lines in the coming week.

We could also fix this in ASPECT, but I think it should go into deal.II.

**Comment by tjhei** [comment]
@bangerth, all this won't solve our issue here though. If you switch to our CG implementation, you will get
```
The iterative (top left) solver in BlockSchurPreconditioner::vmult did not converge. It reported the following error:
Iterative method reported convergence failure in step 10000. The residual in the last step was 0.107138.
```

**Comment by bangerth** [comment]
Well, the change might fix depth_average, not periodic_box. I think these are separate issues.

I don't really know what to do here :-(

**Comment by tjhei** [comment]
Ideas:
1. figure out how to do diagonal strengthening or something for the coarse KLU.
2. give ML the constant mode for the displacement
3. fix the displacement to make the system uniquely solvable (somewhat complicated)

**Comment by bangerth** [comment]
I think you're going to have an extremely hard time to figure out the null space just from the domain and the description of what the boundary conditions are. It's sort of a pattern matching problem that humans are good at but that are difficult to put into rules :-(

**Comment by ian-r-rose** [comment]
For what it's worth, a PETSc build of Aspect runs this test without crashing.

**Comment by gassmoeller** [comment]
closed by #328
